File: app/src/main/java/com/dougkeen/bart/backend/AlertProjection.java
package com.dougkeen.bart.backend;

import com.dougkeen.bart.model.Alert;
import com.dougkeen.bart.networktasks.GtfsRealtimeFeedIndex;
import com.google.transit.realtime.GtfsRealtime;

import java.text.DateFormat;
import java.util.Date;
import java.util.Locale;
import java.util.Objects;

/** Converts the latest alert feed into the app's alert model. */
public final class AlertProjection implements TransitProjection<Alert.AlertList> {
    @Override
    public Alert.AlertList project(TransitFeedSnapshot snapshot) {
        Alert.AlertList result = new Alert.AlertList();
        DateFormat format = DateFormat.getDateTimeInstance(DateFormat.SHORT,
                DateFormat.SHORT, Locale.getDefault());
        GtfsRealtimeFeedIndex index = snapshot.getAlertIndex();
        for (GtfsRealtime.FeedEntity entity : index.getAlertEntities()) {
            if (!entity.hasAlert()) {
                continue;
            }
            GtfsRealtime.Alert source = entity.getAlert();
            Alert alert = new Alert(entity.getId());
            alert.setType(source.hasEffect() ? source.getEffect().name() : "");
            alert.setDescription(text(source.hasHeaderText()
                    ? source.getHeaderText() : null, source.hasDescriptionText()
                    ? source.getDescriptionText() : null));
            if (!source.getActivePeriodList().isEmpty()) {
                GtfsRealtime.TimeRange period = source.getActivePeriod(0);
                if (period.hasStart()) {
                    alert.setPostedTime(format.format(new Date(
                            period.getStart() * 1000L)));
                }
                if (period.hasEnd()) {
                    alert.setExpiresTime(format.format(new Date(
                            period.getEnd() * 1000L)));
                }
            }
            result.addAlert(alert);
        }
        if (!result.hasAlerts()) {
            result.setNoDelaysReported(true);
        }
        return result;
    }

    @Override
    public boolean areEquivalent(Alert.AlertList previous,
                                 Alert.AlertList current) {
        if (previous == current) {
            return true;
        }
        if (previous == null || current == null
                || previous.areNoDelaysReported()
                != current.areNoDelaysReported()
                || previous.getAlerts().size() != current.getAlerts().size()) {
            return false;
        }
        for (int i = 0; i < previous.getAlerts().size(); i++) {
            Alert left = previous.getAlerts().get(i);
            Alert right = current.getAlerts().get(i);
            if (!Objects.equals(left.getId(), right.getId())
                    || !Objects.equals(left.getType(), right.getType())
                    || !Objects.equals(left.getDescription(), right.getDescription())
                    || !Objects.equals(left.getPostedTime(), right.getPostedTime())
                    || !Objects.equals(left.getExpiresTime(), right.getExpiresTime())) {
                return false;
            }
        }
        return true;
    }

    private static String text(GtfsRealtime.TranslatedString header,
                               GtfsRealtime.TranslatedString description) {
        String headerText = translation(header);
        String descriptionText = translation(description);
        if (headerText.isEmpty()) {
            return descriptionText;
        }
        if (descriptionText.isEmpty()) {
            return headerText;
        }
        return headerText + "\n" + descriptionText;
    }

    private static String translation(GtfsRealtime.TranslatedString value) {
        if (value == null || value.getTranslationList().isEmpty()) {
            return "";
        }
        return value.getTranslation(0).getText();
    }
}

File: app/src/main/kotlin/com/dougkeen/bart/backend/AlertProjection.kt
package com.dougkeen.bart.backend

import com.dougkeen.bart.model.Alert
import com.dougkeen.bart.networktasks.GtfsRealtimeFeedIndex
import com.dougkeen.bart.performance.PerformanceTrace
import com.google.transit.realtime.GtfsRealtime

/** Converts the latest alert feed into the app's alert model. */
class AlertProjection {
    fun project(snapshot: TransitFeedSnapshot): Alert.AlertList {
        return PerformanceTrace.section("BART alert projection") {
            val index: GtfsRealtimeFeedIndex = snapshot.getAlertIndex()
            val alerts = mutableListOf<Alert>()
            for (entity in index.alertEntities) {
                if (!entity.hasAlert()) {
                    continue
                }
                val source = entity.alert
                var postedAtMillis: Long? = null
                var expiresAtMillis: Long? = null
                if (source.activePeriodList.isNotEmpty()) {
                    val period = source.getActivePeriod(0)
                    if (period.hasStart()) {
                        postedAtMillis = period.start * 1000L
                    }
                    if (period.hasEnd()) {
                        expiresAtMillis = period.end * 1000L
                    }
                }
                alerts += Alert(
                    id = entity.id,
                    type = if (source.hasEffect()) source.effect.name else "",
                    description = text(
                        if (source.hasHeaderText()) source.headerText else null,
                        if (source.hasDescriptionText()) source.descriptionText else null
                    ),
                    postedAtMillis = postedAtMillis,
                    expiresAtMillis = expiresAtMillis
                )
            }
            Alert.AlertList(alerts, alerts.isEmpty())
        }
    }

    fun areEquivalent(
        previous: Alert.AlertList?,
        current: Alert.AlertList?
    ): Boolean {
        if (previous === current) {
            return true
        }
        if (previous == null || current == null
            || previous.areNoDelaysReported() != current.areNoDelaysReported()
            || previous.getAlerts().size != current.getAlerts().size
        ) {
            return false
        }
        for (index in previous.getAlerts().indices) {
            val left = previous.getAlerts()[index]
            val right = current.getAlerts()[index]
            if (left.id != right.id
                || left.type != right.type
                || left.description != right.description
                || left.postedAtMillis != right.postedAtMillis
                || left.expiresAtMillis != right.expiresAtMillis
            ) {
                return false
            }
        }
        return true
    }

    private fun text(
        header: GtfsRealtime.TranslatedString?,
        description: GtfsRealtime.TranslatedString?
    ): String {
        val headerText = translation(header)
        val descriptionText = translation(description)
        return when {
            headerText.isEmpty() -> descriptionText
            descriptionText.isEmpty() -> headerText
            else -> "$headerText\n$descriptionText"
        }
    }

    private fun translation(value: GtfsRealtime.TranslatedString?): String =
        if (value == null || value.translationList.isEmpty()) ""
        else value.getTranslation(0).text
}
